### ISSUE TITLE
Change group_by so it the key can take any generic value.

### ISSUE THREAD

**Issue by farziengineer** [farziengineer/underscore_cpp#59]
https://github.com/farziengineer/underscore_cpp/blob/2fd26972c8042ae5ed29304e8b072e78807e7df3/underscore.cpp#L313
As of now, it only takes an `int` for the key.

**Comment by Gotham13121997** [comment]
group by return type of the grouping function can be good

**Comment by farziengineer** [comment]
@Gotham13121997 exactly.

**Comment by gubatron** [comment]
think something along the lines of this would fly???

```cpp
template<typename Container, typename Function, typename FunctionReturnType>
	std::map<FunctionReturnType, std::vector<typename Container::value_type> >  group_by(Container &container, Function function) 
	{
std::map<FunctionReturnType, std::vector<typename Container::value_type> > result;
```

**Comment by gubatron** [comment]
when this one is done, then it makes sense to work on #61

**Comment by farziengineer** [comment]
@gubatron want to give it a try, go ahead ☺️

**Comment by gubatron** [comment]
alright, will submit a PR in a couple hours if everything goes smoothly with building and my dev environment.

**Comment by Rigbert** [comment]
I can also take this, if @gubatron doesn't have time for it.

**Comment by farziengineer** [comment]
Hi @Rigbert, I think @gubatron might need more time. @gubatron please keep us updated.

**Comment by gubatron** [comment]
I got something working already, but I'm not happy with it, iterating over it some more.

It looks like this now:

Usage:
```cpp
// takes a value as the way for the compiler to be hinted on what the key type is
std::map<bool, std::vector<int> > mp = _::group_by(vec, is_odd, false);
std::map<bool, std::vector<int> > mp = _::group_by(vec, is_odd, false);
std::map<int, std::vector<int> > mp = _::group_by(vec, is_odd, 1337);
```

The signature looks like this:
```cpp
template <typename Container, typename Function, typename KeyType>
auto group_by(Container &container, Function function, KeyType keyHint) -> std::map< KeyType, std::vector<typename Container::value_type> >;
```

I'm working with auto because I was playing with `decltype` as a way to get what the `KeyType` would be.

My goal is to have it work like this, without passing values as type hints to the compiler
```cpp
std::map<int, std::vector<int> > mp = _::group_by<int>(vec, is_odd);
```

**Comment by gubatron** [comment]
It also works like this, without `auto`:
```cpp
template <typename Container, typename Function, typename KeyType>
std::map< KeyType, std::vector<typename Container::value_type> > group_by(Container &container, Function function, KeyType keyHint);
```

but again, not happy until I get the usage to allow me to specify the type itself, but dealing with lots of errors

**Comment by gubatron** [comment]
damn, already a couple of answers

**Comment by farziengineer** [comment]
Damn 😀

**Comment by gubatron** [comment]
got it working like this:
`std::map<int, std::vector<int> > mp = _::group_by<int>(vec, is_odd);`

but now I'm gonna make it even more magic so it works like this:
`std::map<int, std::vector<int> > mp = _::group_by(vec, is_odd);`

using `decltype` and `auto`, my issue was the order of the template parameters, good lesson learned.

**Comment by gubatron** [comment]
There you go guys, sorry for the delay
https://github.com/farziengineer/underscore_cpp/pull/64

**Comment by Gotham13121997** [comment]
closed via #64

**Comment by farziengineer** [comment]
Thanks a lot @gubatron. Learnt a lot here.
